### PR TITLE
Introduced COUApplication, COUUnit and COUMachine

### PR DESCRIPTION
**Pull request by rgildein**
Add COUApplication, COUUnit and COUMachine to juju_utils and create ger_applications function, which is obtaining all application from model with all information required for COU.

### CODE REVIEW

File: cou/utils/juju_utils.py [hunk 1]
```diff
@@ -266,6 +313,38 @@ async def connect(self) -> None:
...
+        # note(rgildein): We get the applications from the Juju status, since we can get more
+        #                 information the status than from objects. e.g. workload_version for unit
+        full_status = await self.get_status()
+        machines = await self.get_model_machines()
+
+        return {
+            app: COUApplication(
+                name=app,
+                can_upgrade_to=status.can_upgrade_to,
+                charm=status.charm,
```
Comment (rgildein) [inline cou/utils/juju_utils.py:331]: This is not what we need, value like this look like `'ch:amd64/focal/ubuntu-24'` and it should be just `ubuntu`.

File: cou/utils/juju_utils.py [hunk 2]
```diff
@@ -266,6 +313,38 @@ async def connect(self) -> None:
...
+        """Return list of applications with all relevant information.
+
+        :returns: list of application with all information
+        :rtype: list[COUApplication]
+        """
+        # note(rgildein): We get the applications from the Juju status, since we can get more
+        #                 information the status than from objects. e.g. workload_version for unit
+        full_status = await self.get_status()
+        machines = await self.get_model_machines()
+
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:326]: I think you can also get the applications like: `model.applications` and with that access the charm_name and the config without the need of using `self.get_application_config`

**Review by gabrielcocenza — CHANGES_REQUESTED** [review]
Small details, almost ready for merging

File: cou/utils/juju_utils.py [hunk 3]
```diff
@@ -266,6 +320,39 @@ async def connect(self) -> None:
         )
 
     @retry
+    async def get_applications(self) -> dict[str, COUApplication]:
```
[the remark this answers is not in the document]
Comment (rgildein) [reply cou/utils/juju_utils.py:323]: It's COUModel, saying that it returns cou_application is redundant.

File: cou/utils/juju_utils.py [hunk 4]
```diff
@@ -266,6 +320,39 @@ async def connect(self) -> None:
...
+            app: COUApplication(
+                name=app,
+                can_upgrade_to=status.can_upgrade_to,
+                charm=model.applications[app].charm_name,
+                charm_channel=status.charm_channel,
+                config=await model.applications[app].get_config(),
+                machines={unit.machine: machines[unit.machine] for unit in status.units.values()},
+                model=self,
+                series=status.series,
+                subordinate_to=status.subordinate_to,
```
[the remark this answers is not in the document]
Comment (rgildein) [reply cou/utils/juju_utils.py:345]: We are not using it, that's true, but I see benefit leaving it there. I'll move the property `is_subordinate` to the COUApplication, what do you say?

File: cou/utils/juju_utils.py [hunk 5]
```diff
@@ -131,6 +133,67 @@ async def wrapper(*args: Any, **kwargs: Any) -> Any:  # pylint: disable=W9011
...
+
+
+@dataclass(frozen=True)
+class COUApplication:
+    """Representation of an single application."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    name: str
+    can_upgrade_to: str
```
Comment (agileshaw) [inline cou/utils/juju_utils.py:177]: Question: why is this attribute needed?

**Review by agileshaw — COMMENTED** [review]
Overall looks good. Two small comments inline.

File: cou/utils/juju_utils.py (hunk 5 above)
Comment (rgildein) [reply cou/utils/juju_utils.py:177]: It is defined in `OpenStackApplication`, but never used. So it's not needed.

Comment (rgildein) [reply cou/utils/juju_utils.py:177]: Should I remove it?

Comment (agileshaw) [reply cou/utils/juju_utils.py:177]: Just checked, it is being used to determine whether we are at the latest revision of the current channel ([line](https://github.com/canonical/charmed-openstack-upgrader/blob/bc6fe5d89f784d0a8e57056fea9e1297bd9fb8d3/cou/apps/base.py#L392)). So we should leave it.

**Review by agileshaw — APPROVED** [review]
LGTM

File: cou/utils/juju_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rgildein) [reply cou/utils/juju_utils.py:365]: I removed it in the #242, where the real implementation happen.

**Pull request merged**